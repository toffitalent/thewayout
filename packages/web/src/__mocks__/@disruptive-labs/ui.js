const actual = jest.requireActual('@disruptive-labs/ui');

const showToast = jest.fn();

actual.Toast.show = showToast;
actual.Toast.danger = showToast;
actual.Toast.info = showToast;
actual.Toast.success = showToast;
actual.Toast.warning = showToast;

module.exports = {
  ...actual,
  showToast,
};
