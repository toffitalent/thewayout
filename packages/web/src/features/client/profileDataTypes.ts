import {
  FieldComponentProps,
  FieldType,
  WizardChoicesProps,
  WizardStep,
} from '@disruptive-labs/ui';
import {
  Age,
  Education,
  Ethnicity,
  ExperienceSkills,
  Gender,
  JusticeStatus,
  Languages,
  License,
  MaritalStatus,
  OffenseWizardType,
  Orientation,
  PersonalStrengths,
  ReferredBy,
  RelativeExperience,
  ReleasedAt,
  Religion,
  State,
  StateFederal,
  Support,
  TimeServed,
  VeteranBranchOfService,
  VeteranCampaign,
  VeteranOrJustice,
  VeteranRank,
  VeteranStatus,
  VeteranTypeDischarge,
} from '@two/shared';

export enum StepsClient {
  veteranOrJustice = 'veteranOrJustice',

  veteranService = 'veteranService',
  veteranReservist = 'veteranReservist',
  veteranCampaigns = 'veteranCampaigns',
  veteranDates = 'veteranDates',
  veteranTypeDischarge = 'veteranTypeDischarge',
  veteranRank = 'veteranRank',
  veteranDd214 = 'veteranDd214',

  name = 'name',
  justice = 'justice',
  facility = 'facility',
  expectedReleasedAt = 'expectedReleasedAt',
  releasedCounty = 'releasedCounty',
  phone = 'phone',
  address = 'address',

  offenseCategory = 'offenseCategory',
  sexualOffenderRegistry = 'sexualOffenderRegistry',
  sbn = 'sbn',
  timeServed = 'timeServed',
  releasedAt = 'releasedAt',
  stateOrFederal = 'stateOrFederal',

  support = 'support',
  reentryServiceProvider = 'reentryServiceProvider',
  reentryPipeline = 'reentryPipeline',
  personalStrengths = 'personalStrengths',
  experienceSkills = 'experienceSkills',
  languages = 'languages',

  relativeExperience = 'relativeExperience',
  personalSummary = 'personalSummary',
  education = 'education',
  license = 'license',
  diversityInclusion = 'diversityInclusion',
  referredBy = 'referredBy',
}

export interface WizardData extends OffenseWizardType {
  veteranOrJustice: VeteranOrJustice[];

  firstName: string;
  lastName: string;
  phone?: string;
  // address
  address: string;
  city: string;
  state: State | string;
  postalCode: string;

  support: Support[];
  rspId?: string;
  isNewRspMember?: boolean;

  // diversityInclusion
  gender?: Gender;
  orientation?: Orientation;
  religion?: Religion;
  maritalStatus?: MaritalStatus;
  age?: Age;
  disability?: keyof typeof booleanOption;
  ethnicity?: Ethnicity;
  veteranStatus?: VeteranStatus;

  referredBy: ReferredBy;
  personalStrengths: PersonalStrengths[];
  experience: ExperienceSkills[];
  languages: Languages[];
  relativeExperience: (RelativeExperience & { stillWork: boolean })[];
  personalSummary: string;
  education: Education[];
  license: License[];

  // only justiceImpacted clients
  justiceStatus?: JusticeStatus;
  sexualOffenderRegistry?: keyof typeof booleanOption | null;
  sbn?: keyof typeof booleanOption | null;
  timeServed?: TimeServed;
  releasedAt?: ReleasedAt;
  stateOrFederal?: StateFederal;
  facility: string;
  expectedReleasedAt?: string;
  releasedCounty?: string;

  // only veteran clients
  veteranService?: VeteranBranchOfService[];
  veteranCampaigns?: VeteranCampaign[];
  veteranRank?: VeteranRank;
  veteranStartAtMonth?: string;
  veteranStartAtYear?: string;
  veteranEndAtMonth?: string;
  veteranEndAtYear?: string;
  veteranReservist?: keyof typeof booleanOption;
  veteranTypeDischarge?: VeteranTypeDischarge;
  veteranDd214?: keyof typeof booleanOption;
}

export const booleanOption = {
  yes: true,
  no: false,
};

type WizardFormFields<T> = T extends FieldType ? Omit<FieldComponentProps<T>, 'control'> : never;
export type CustomWizardField = (
  | WizardFormFields<FieldType>
  | (Omit<WizardChoicesProps, 'control'> & {
      type: 'choices';
    })
) & {
  conditionalProps?: (
    fields: Partial<
      WizardData & RelativeExperience & Education & License & Languages & { stillWork: boolean }
    >,
  ) => { [key: string]: any };
  conditionalPropsParams?: string[];
};

export type CustomWizardStep = Omit<WizardStep, 'fields' | 'id'> & {
  id: StepsClient;
  fields?: CustomWizardField[];
};

type FilterConditionally<Source, Condition> = Pick<
  Source,
  { [K in keyof Source]-?: Source[K] extends Condition ? K : never }[keyof Source]
>;

export type ArrayStepWizardData = FilterConditionally<WizardData, object[]>;
export type ArrayStepItem = ArrayStepWizardData[keyof ArrayStepWizardData][0];
export type ArrayStepItems = Array<ArrayStepItem>;

type KeysOfUnion<T> = T extends T ? keyof T : never;
export type AvailableKeys = KeysOfUnion<ArrayStepItem>;
