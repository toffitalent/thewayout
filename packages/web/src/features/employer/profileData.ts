import { WizardStep } from '@disruptive-labs/ui';
import { Industry, NumberOfEmployers, YearsInBusiness } from '@two/shared';
import { states } from '@two/shared/src/constants/states';

export enum Steps {
  name = 'name',
  industry = 'industry',
  description = 'description',
  details = 'details',
  location = 'location',
}

export const stepsConfig: Record<Steps, { id: string; title: string }> = {
  [Steps.name]: { id: 'name', title: 'Employer Name' },
  [Steps.industry]: { id: 'industry', title: 'Industry' },
  [Steps.description]: { id: 'description', title: 'Company Description' },
  [Steps.details]: { id: 'details', title: 'Company Details' },
  [Steps.location]: { id: 'location', title: 'Company Location' },
};

export const steps: WizardStep[] = [
  {
    ...stepsConfig[Steps.name],
    fields: [
      {
        label: 'Name',
        name: 'name',
        type: 'text',
        placeholder: 'E.g. ABC Company',
        validate: { required: true },
      },
    ],
    nextButtonLabel: 'Save & Next',
    nextStep: 'industry',
  },
  {
    ...stepsConfig[Steps.industry],
    fields: [
      {
        label: 'Industry',
        name: 'industry',
        type: 'select',
        options: Object.values(Industry).map((el) => ({ label: el, value: el })),
        placeholder: 'Select',
        validate: { required: true },
      },
    ],
    nextButtonLabel: 'Save & Next',
    showBackButton: true,
    nextStep: 'description',
  },
  {
    ...stepsConfig[Steps.description],
    fields: [
      {
        label: 'Description',
        name: 'description',
        type: 'text',
        multiline: true,
        placeholder: 'Your Company description... ',
        validate: { required: true, maxLength: 1000 },
        helperText: 'Max 1,000 characters.',
      },
    ],
    nextButtonLabel: 'Save & Next',
    showBackButton: true,
    nextStep: 'details',
  },
  {
    ...stepsConfig[Steps.details],
    fields: [
      {
        label: 'Years in Business',
        name: 'yearsInBusiness',
        type: 'select',
        options: Object.values(YearsInBusiness).map((el) => ({ label: `${el} years`, value: el })),
        placeholder: 'Select',
        validate: { required: true },
      },
      {
        label: 'Number of Employees',
        name: 'numberOfEmployees',
        type: 'select',
        options: Object.values(NumberOfEmployers).map((el) => ({ label: el, value: el })),
        placeholder: 'Select',
        validate: { required: true },
      },
    ],
    nextButtonLabel: 'Save & Next',
    showBackButton: true,
    nextStep: 'location',
  },
  {
    ...stepsConfig[Steps.location],
    fields: [
      {
        label: 'Address',
        name: 'address',
        type: 'text',
        placeholder: 'E.g. 459 S Hewitt St',
        validate: { required: true },
      },
      {
        label: 'City',
        name: 'city',
        type: 'text',
        placeholder: 'E.g. Los Angeles',
        validate: { required: true },
      },
      {
        label: 'State',
        name: 'state',
        type: 'select',
        options: Object.entries(states).map(([key, value]) => ({ value: key, label: value })),
        placeholder: 'Select',
        validate: { required: true },
      },
      {
        label: 'ZIP Code',
        name: 'postalCode',
        type: 'text',
        placeholder: 'E.g. 90255',
        validate: { required: true },
      },
    ],
    nextButtonLabel: 'Save & Next',
    showBackButton: true,
    nextStep: 'confirm',
  },
];
