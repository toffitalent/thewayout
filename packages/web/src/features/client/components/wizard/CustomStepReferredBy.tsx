import {
  Button,
  FieldOptions,
  useForm,
  useWizardContext,
  WizardChoices,
  WizardChoicesGroup,
  WizardStep,
} from '@disruptive-labs/ui';
import ArrowNarrowRight from '@disruptive-labs/ui/dist/icons/ArrowNarrowRight';
import { selectAuthUser } from '@app/features/auth';
import { useAppSelector } from '@app/hooks';
import { WizardData } from '../../profileDataTypes';

export interface CustomStepReferredByProps extends Omit<Partial<WizardStep>, 'fields'> {
  fields?: {
    name: string;
    type: 'choices';
    label: string;
    placeholder: string;
    options: (FieldOptions | WizardChoicesGroup)[];
  }[];
}

export const CustomStepReferredBy = (props: CustomStepReferredByProps) => {
  const user = useAppSelector(selectAuthUser);
  const { next } = useWizardContext<WizardData>();

  const {
    submitForm,
    control,
    formState: { isDirty, isSubmitting },
  } = useForm({
    defaultValues: {
      referredBy: user?.client?.referredBy,
    },
    onSubmit: async ({ referredBy }) => {
      await next({ referredBy });
    },
  });

  const { fields } = props;
  if (!fields) {
    return null;
  }

  return (
    <form onSubmit={submitForm}>
      {fields.map((field) => (
        <WizardChoices key={field.name} {...field} control={control} />
      ))}
      <Button
        colorScheme="primary"
        fluid
        type="submit"
        maxWidth={96}
        mt={8}
        accessoryRight={<ArrowNarrowRight />}
        disabled={(!user?.client?.referredBy && !isDirty) || isSubmitting}
      >
        Submit
      </Button>
    </form>
  );
};
