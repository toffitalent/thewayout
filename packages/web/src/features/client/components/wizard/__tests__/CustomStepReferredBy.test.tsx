import { useWizardContext } from '@disruptive-labs/ui';
import userEvent from '@testing-library/user-event';
import { render, waitFor } from '@test';
import { steps } from '../../../profileData';
import { StepsClient } from '../../../profileDataTypes';
import { CustomStepReferredBy, CustomStepReferredByProps } from '../CustomStepReferredBy';

jest.mock('@disruptive-labs/ui', () => ({
  ...jest.requireActual('@disruptive-labs/ui'),
  useWizardContext: jest.fn(),
}));
const referredByStep = steps.find((el) => el.id === StepsClient.referredBy);

describe('CustomStepReferredBy', () => {
  beforeEach(() => {
    (useWizardContext as any).mockImplementation(() => ({
      data: {},
      goTo: jest.fn(),
    }));
  });

  test('renders correctly', async () => {
    const { container, findByText } = render(
      <CustomStepReferredBy fields={(referredByStep as CustomStepReferredByProps).fields || []} />,
    );
    await waitFor(async () =>
      expect(await findByText('Case Manager / Reentry Service Provider')).toBeInTheDocument(),
    );
    expect(container).toMatchSnapshot();
  });

  test('press next & save button', async () => {
    const next = jest.fn();
    (useWizardContext as any).mockImplementation(() => ({
      data: {
        referredBy: '',
      },
      next,
    }));
    const { container, getByText } = render(
      <CustomStepReferredBy {...(referredByStep as CustomStepReferredByProps)} />,
    );
    await userEvent.click(getByText('Newsletter'));
    await userEvent.click(getByText('Submit'));
    expect(next).toBeCalledWith({ referredBy: 'newsletter' });
    expect(container).toMatchSnapshot();
  });
});
