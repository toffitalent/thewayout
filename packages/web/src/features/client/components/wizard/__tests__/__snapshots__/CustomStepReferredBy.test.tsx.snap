// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CustomStepReferredBy press next & save button 1`] = `
<div>
  <form>
    <div
      class="control control fluid"
      role="group"
    >
      <fieldset
        class="choices inline"
      >
        <div
          role="radiogroup"
        >
          <div
            aria-checked="false"
            aria-disabled="false"
            class="choicesItem"
            role="radio"
            tabindex="0"
          >
            <span
              class="choicesItemText"
            >
              Case Manager / Reentry Service Provider
            </span>
            <span
              class="choicesItemIcon"
            >
              <svg
                focusable="false"
                role="presentation"
                viewBox="0 0 24 24"
              >
                <polygon
                  fill="currentColor"
                  points="9.3,19 3,12.8 5.3,10.4 9.3,14.3 18.7,5 21,7.3 "
                />
              </svg>
            </span>
          </div>
          <div
            aria-checked="false"
            aria-disabled="false"
            class="choicesItem"
            role="radio"
            tabindex="0"
          >
            <span
              class="choicesItemText"
            >
              Probation / Parole Officer / Social Worker
            </span>
            <span
              class="choicesItemIcon"
            >
              <svg
                focusable="false"
                role="presentation"
                viewBox="0 0 24 24"
              >
                <polygon
                  fill="currentColor"
                  points="9.3,19 3,12.8 5.3,10.4 9.3,14.3 18.7,5 21,7.3 "
                />
              </svg>
            </span>
          </div>
          <div
            aria-checked="true"
            aria-disabled="false"
            class="choicesItem"
            role="radio"
            tabindex="0"
          >
            <span
              class="choicesItemText"
            >
              Newsletter
            </span>
            <span
              class="choicesItemIcon"
            >
              <svg
                focusable="false"
                role="presentation"
                viewBox="0 0 24 24"
              >
                <polygon
                  fill="currentColor"
                  points="9.3,19 3,12.8 5.3,10.4 9.3,14.3 18.7,5 21,7.3 "
                />
              </svg>
            </span>
          </div>
          <div
            aria-checked="false"
            aria-disabled="false"
            class="choicesItem"
            role="radio"
            tabindex="0"
          >
            <span
              class="choicesItemText"
            >
              Social media
            </span>
            <span
              class="choicesItemIcon"
            >
              <svg
                focusable="false"
                role="presentation"
                viewBox="0 0 24 24"
              >
                <polygon
                  fill="currentColor"
                  points="9.3,19 3,12.8 5.3,10.4 9.3,14.3 18.7,5 21,7.3 "
                />
              </svg>
            </span>
          </div>
          <div
            aria-checked="false"
            aria-disabled="false"
            class="choicesItem"
            role="radio"
            tabindex="0"
          >
            <span
              class="choicesItemText"
            >
              News / Media
            </span>
            <span
              class="choicesItemIcon"
            >
              <svg
                focusable="false"
                role="presentation"
                viewBox="0 0 24 24"
              >
                <polygon
                  fill="currentColor"
                  points="9.3,19 3,12.8 5.3,10.4 9.3,14.3 18.7,5 21,7.3 "
                />
              </svg>
            </span>
          </div>
          <div
            aria-checked="false"
            aria-disabled="false"
            class="choicesItem"
            role="radio"
            tabindex="0"
          >
            <span
              class="choicesItemText"
            >
              Family / Friends
            </span>
            <span
              class="choicesItemIcon"
            >
              <svg
                focusable="false"
                role="presentation"
                viewBox="0 0 24 24"
              >
                <polygon
                  fill="currentColor"
                  points="9.3,19 3,12.8 5.3,10.4 9.3,14.3 18.7,5 21,7.3 "
                />
              </svg>
            </span>
          </div>
          <div
            aria-checked="false"
            aria-disabled="false"
            class="choicesItem"
            role="radio"
            tabindex="0"
          >
            <span
              class="choicesItemText"
            >
              Other
            </span>
            <span
              class="choicesItemIcon"
            >
              <svg
                focusable="false"
                role="presentation"
                viewBox="0 0 24 24"
              >
                <polygon
                  fill="currentColor"
                  points="9.3,19 3,12.8 5.3,10.4 9.3,14.3 18.7,5 21,7.3 "
                />
              </svg>
            </span>
          </div>
        </div>
      </fieldset>
    </div>
    <button
      class="button md solid c-primary fluid maxWidth-96 marginTop-8"
      type="submit"
    >
      Submit
      <span
        aria-hidden="true"
        class="accessoryRight"
      >
        <svg
          aria-hidden="true"
          class="icon"
          focusable="false"
          stroke-width="2"
          viewBox="0 0 24 24"
        >
          <path
            d="M0 0h24v24H0z"
            stroke="none"
          />
          <path
            d="M5 12h14M15 16l4-4M15 8l4 4"
          />
        </svg>
      </span>
    </button>
  </form>
</div>
`;

exports[`CustomStepReferredBy renders correctly 1`] = `
<div>
  <form>
    <div
      class="control control fluid"
      role="group"
    >
      <fieldset
        class="choices inline"
      >
        <div
          role="radiogroup"
        >
          <div
            aria-checked="false"
            aria-disabled="false"
            class="choicesItem"
            role="radio"
            tabindex="0"
          >
            <span
              class="choicesItemText"
            >
              Case Manager / Reentry Service Provider
            </span>
            <span
              class="choicesItemIcon"
            >
              <svg
                focusable="false"
                role="presentation"
                viewBox="0 0 24 24"
              >
                <polygon
                  fill="currentColor"
                  points="9.3,19 3,12.8 5.3,10.4 9.3,14.3 18.7,5 21,7.3 "
                />
              </svg>
            </span>
          </div>
          <div
            aria-checked="false"
            aria-disabled="false"
            class="choicesItem"
            role="radio"
            tabindex="0"
          >
            <span
              class="choicesItemText"
            >
              Probation / Parole Officer / Social Worker
            </span>
            <span
              class="choicesItemIcon"
            >
              <svg
                focusable="false"
                role="presentation"
                viewBox="0 0 24 24"
              >
                <polygon
                  fill="currentColor"
                  points="9.3,19 3,12.8 5.3,10.4 9.3,14.3 18.7,5 21,7.3 "
                />
              </svg>
            </span>
          </div>
          <div
            aria-checked="false"
            aria-disabled="false"
            class="choicesItem"
            role="radio"
            tabindex="0"
          >
            <span
              class="choicesItemText"
            >
              Newsletter
            </span>
            <span
              class="choicesItemIcon"
            >
              <svg
                focusable="false"
                role="presentation"
                viewBox="0 0 24 24"
              >
                <polygon
                  fill="currentColor"
                  points="9.3,19 3,12.8 5.3,10.4 9.3,14.3 18.7,5 21,7.3 "
                />
              </svg>
            </span>
          </div>
          <div
            aria-checked="false"
            aria-disabled="false"
            class="choicesItem"
            role="radio"
            tabindex="0"
          >
            <span
              class="choicesItemText"
            >
              Social media
            </span>
            <span
              class="choicesItemIcon"
            >
              <svg
                focusable="false"
                role="presentation"
                viewBox="0 0 24 24"
              >
                <polygon
                  fill="currentColor"
                  points="9.3,19 3,12.8 5.3,10.4 9.3,14.3 18.7,5 21,7.3 "
                />
              </svg>
            </span>
          </div>
          <div
            aria-checked="false"
            aria-disabled="false"
            class="choicesItem"
            role="radio"
            tabindex="0"
          >
            <span
              class="choicesItemText"
            >
              News / Media
            </span>
            <span
              class="choicesItemIcon"
            >
              <svg
                focusable="false"
                role="presentation"
                viewBox="0 0 24 24"
              >
                <polygon
                  fill="currentColor"
                  points="9.3,19 3,12.8 5.3,10.4 9.3,14.3 18.7,5 21,7.3 "
                />
              </svg>
            </span>
          </div>
          <div
            aria-checked="true"
            aria-disabled="false"
            class="choicesItem"
            role="radio"
            tabindex="0"
          >
            <span
              class="choicesItemText"
            >
              Family / Friends
            </span>
            <span
              class="choicesItemIcon"
            >
              <svg
                focusable="false"
                role="presentation"
                viewBox="0 0 24 24"
              >
                <polygon
                  fill="currentColor"
                  points="9.3,19 3,12.8 5.3,10.4 9.3,14.3 18.7,5 21,7.3 "
                />
              </svg>
            </span>
          </div>
          <div
            aria-checked="false"
            aria-disabled="false"
            class="choicesItem"
            role="radio"
            tabindex="0"
          >
            <span
              class="choicesItemText"
            >
              Other
            </span>
            <span
              class="choicesItemIcon"
            >
              <svg
                focusable="false"
                role="presentation"
                viewBox="0 0 24 24"
              >
                <polygon
                  fill="currentColor"
                  points="9.3,19 3,12.8 5.3,10.4 9.3,14.3 18.7,5 21,7.3 "
                />
              </svg>
            </span>
          </div>
        </div>
      </fieldset>
    </div>
    <button
      class="button md solid c-primary fluid maxWidth-96 marginTop-8"
      type="submit"
    >
      Submit
      <span
        aria-hidden="true"
        class="accessoryRight"
      >
        <svg
          aria-hidden="true"
          class="icon"
          focusable="false"
          stroke-width="2"
          viewBox="0 0 24 24"
        >
          <path
            d="M0 0h24v24H0z"
            stroke="none"
          />
          <path
            d="M5 12h14M15 16l4-4M15 8l4 4"
          />
        </svg>
      </span>
    </button>
  </form>
</div>
`;
